Mathlib/Analysis/Convex/Cone/Caratheodory.lean

```diff
@@ -0,0 +1,249 @@
+          exact Subtype.val_injective
+        · -- At least one coefficient is 0.
+          have : k d = 0 := by
+            rw [Nonneg.mk_eq_zero, div_mul_cancel, sub_self]
+            exact (ne_of_lt hd₁.2).symm
+          rw [this, zero_smul]
+
+variable {s : Set E} {x : E} (hx : x ∈ toPointedCone 𝕜 s)
+
+/-- Given a point `x` in the convex cone of a set `s`, this is a finite subset of `s` of minimum
+cardinality, whose convex cone contains `x`. -/
+noncomputable def minCardFinsetOfMemtoPointedCone (hx : x ∈ s.toPointedCone 𝕜) : Finset E :=
+  Function.argminOn Finset.card Nat.lt_wfRel.2 { t | ↑t ⊆ s ∧ x ∈ (t : Set E).toPointedCone 𝕜 }
+    <| by exact Submodule.mem_span_finite_of_mem_span hx
+
+theorem minCardFinsetOftoPointedCone_subseteq : ↑(minCardFinsetOfMemtoPointedCone hx) ⊆ s :=
+  (Function.argminOn_mem _ _ { t : Finset E | ↑t ⊆ s ∧ x ∈ (t : Set E).toPointedCone 𝕜 } _).1
+
+theorem mem_minCardFinsetOfMemtoPointedCone :
+    x ∈ (minCardFinsetOfMemtoPointedCone hx : Set E).toPointedCone 𝕜  := by
+  have hs : Set.Nonempty {(t : Finset E) | (t : Set E) ⊆ s ∧ x ∈ toPointedCone 𝕜 ↑t} := by
+    exact Submodule.mem_span_finite_of_mem_span hx
+  have h := (Function.argminOn_mem Finset.card Nat.lt_wfRel.2
+    { t : Finset E | ↑t ⊆ s ∧ x ∈ (t : Set E).toPointedCone 𝕜 } hs).2
+  -- deterministic timeout if we use `exact` directly instead of the intermediate `have`
+  exact h
+
+theorem minCardFinsetOfMemtoPointedCone_card_le_card {t : Finset E} (ht₁ : ↑t ⊆ s)
+    (ht₂ : x ∈ (t : Set E).toPointedCone 𝕜) : (minCardFinsetOfMemtoPointedCone hx).card ≤ t.card :=
+  Function.argminOn_le _ _ _ (by exact ⟨ht₁, ht₂⟩)
+
+theorem linearIndependent_minCardFinsetOfMemtoPointedCone :
+    LinearIndependent 𝕜 ((↑) : minCardFinsetOfMemtoPointedCone hx → E) := by
+  by_cases h : minCardFinsetOfMemtoPointedCone hx = ∅
+  · rw [h]
+    exact linearIndependent_empty_type
+  · have : 0 < (minCardFinsetOfMemtoPointedCone hx).card := by
+      rw [card_pos]
+      exact nonempty_of_ne_empty h
+    set k := (minCardFinsetOfMemtoPointedCone hx).card - 1 with hk
+    have hk : (minCardFinsetOfMemtoPointedCone hx).card = k + 1 := by
+      rwa [hk, ← Nat.succ_eq_add_one, ← Nat.pred_eq_sub_one, Nat.succ_pred_eq_of_pos]
+    classical
+    by_contra h
+    obtain ⟨p, hp⟩ := mem_toPointedCone_erase h (mem_minCardFinsetOfMemtoPointedCone hx)
+    have contra := minCardFinsetOfMemtoPointedCone_card_le_card hx (Set.Subset.trans
+      (Finset.erase_subset (p : E) (minCardFinsetOfMemtoPointedCone hx))
+      (minCardFinsetOftoPointedCone_subseteq hx)) hp
+    rw [← not_lt] at contra
+    apply contra
+    erw [card_erase_of_mem p.2, hk]
+    exact lt_add_one _
+
+end Caratheodory
+
+variable {s : Set E}
+
+/-- **Carathéodory's convexity theorem for convex cones** -/
+
+theorem toPointedCone_eq_union : (s.toPointedCone 𝕜 : Set E) =
+    ⋃ (t : Finset E) (_ : ↑t ⊆ s) (_ : LinearIndependent 𝕜 ((↑) : t → E)),
+      (SetLike.coe $ toPointedCone 𝕜 t) := by
```

Comment:
I have the impression that in lean4, the notation $ has been replaced by <| …

Reply:
Oops. Fixed, thanks!